fixes dotnet/templating#4099  ordering of search results based on download count

vlada-shubina opened the pull request:
### Problem
#4099

### Solution
Ordering of search results now based on download count (descending) and then by name.
Note: for ordering displayed download count (estimated to thousands) is used (not actual).

This means for: A with 1001 downloads and B with 1999 downloads will be displayed as
```
Z    10k
A     1k
B     1k
C    <1k
```

This fix should be included to 6.0.2xx.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

bekir-ozturk commented on src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:
```diff
@@ -205,13 +206,18 @@ private static bool ValidateCommandInput(INewCommandInput commandInput)
         /// </summary>
         private class SearchResultTableRow
         {
+            private const string MinimumDownloadCount = "<1k";
+            private const char ThousandsChar = 'k';
```
These should likely be localized, as we write `b` in Turkish instead of `k`.

vlada-shubina replied:
that's a good point, however it opens other questions, I'm not sure it always follows the number in other languages.
Will 'k' be misunderstood in Turkish? In Russian, it's also different, but people knows 'k' as well.

vlada-shubina replied:
Also I would like to solve it in different PR if needed, it's not related to ordering which has to be fixed in 6.0.2xx

bekir-ozturk replied:
> I'm not sure it always follows the number in other languages.

We can maybe solve this by formatting the string like "{0}k" where `{0}` will be replaced by the number.

vlada-shubina replied:
adding @baronfel here. @baronfel do you think the 'k' part in download count should be localized? Initial decision was not to localize it.

baronfel replied:
There's a wrinkle here when it comes to supporting localization for Japanese numbers (and Japanese localization is something we already support, and so should continue doing so here). Here's a [Unicode language reference on number patterns](https://cldr.unicode.org/translation/number-currency-formats/number-and-currency-patterns#h.eradhhuxzqqz) for context, embedded below:

<html><body>
<!--StartFragment-->

Code | Meaning | English                  Pattern | English                  Formatted                  Example | Japanese                  Pattern | Japanese                  Formatted                  Example
-- | -- | -- | -- | -- | --
4-digit-short-one                  (1000) | 1 thousand = 1,000 | 0K | 1K | 0 (see translation tip below) | 1
5-digit-short-one                  (10000) | 10,000 | 00K | 12K | 0万 | 1万
6-digit-short-one (100000) | 100,000 | 000K | 123K | 00万 | 12万
7-digit-short-one                  (1000000) | 1,000,000 | 0.0M | 1.2M | 000万 | 123万
8-digit-short-one                  (10000000) | 10,000,000 | 00M | 12M | 0000万 | 1234万
9-digit-short-one                  (100000000) | 100,000,000 | 000M | 123M | 0億 | 1           億
10-digit-short-one (1000000000) | 1,000,000,000 | 0B | 1B | 00億 | 12億
11-digit-short-one (10000000000) | 10,000,000,000 | 00B | 12B | 000億 | 123億
12-digit-short-one (100000000000) | 100,000,000,000 | 000B | 123B | 0000億 | 1234億
13-digit-short-one                  (1000000000000) | 1,000,000,000,000 | 0T | 1T | 0兆 | 1兆
14-digit-short-one                  (10000000000000) | 10,000,000,000,000 | 00T | 12T | 00兆 | 12兆
15-digit-short-one                  (100000000000000) | 100,000,000,000,000 | 000T | 123T | 000兆 | 123兆

<!--EndFragment-->
</body>
</html>

The gist is that Japanese has a specific concept for 10k units, and I don't know how well .NET formatting/our localization handles that.

vlada-shubina replied:
https://github.com/dotnet/templating/issues/4223

bekir-ozturk commented on src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:
```diff
@@ -224,18 +230,51 @@ internal string PrintableTotalDownloads
+
+                    if (x != MinimumDownloadCount)
+                    {
+                        int.TryParse(x.Substring(0, x.Length - 1), out xInt);
```
Why `-1`? Is it length of the word `k`? If yes, I prefer `ThousandsChar.Length`.

vlada-shubina replied:
`ThousandsChar` is a char. not a string.

bekir-ozturk replied:
Ah, that's fair. I actually had the assumption in mind that this would be localized, in which case it could be more than 1 character. If you convert the thousands symbol to a `string` for loc purposes, please update this as well.

Otherwise, maybe a comment could help future devs.
```suggestion
                        // x.Length - 1 : remove last character which is always 'k'.
                        int.TryParse(x.Substring(0, x.Length - 1), out xInt);
```

vlada-shubina replied:
depends on decision made in https://github.com/dotnet/templating/pull/4189#discussion_r766764265

vlada-shubina replied:
not relevant anymore

bekir-ozturk commented on src/Microsoft.TemplateEngine.Cli/TemplateSearch/CliTemplateSearchCoordinator.cs:
```diff
@@ -224,18 +230,51 @@ internal string PrintableTotalDownloads
+                    }
+                    if (y != MinimumDownloadCount)
+                    {
+                        int.TryParse(y.Substring(0, y.Length - 1), out yInt);
```
I wish we could store row data in a type safe way instead of parsing it back.

vlada-shubina replied:
We cannot - current formatter only works with strings. The alternative is to sort the collection before rendering.

bekir-ozturk commented on test/dotnet-new3.UnitTests/DotnetNewSearch.cs:
```diff
@@ -877,5 +885,36 @@ public int Compare(string? left, string? right)
             }
         }
+
+        private class DownloadCountComparer : IComparer<string>
```
When we change the implementation in CLI project in the future, will there be some kind of reminder to tell us that we need to change this class as well?

vlada-shubina replied:
Then it would be a valid test failure :)

bekir-ozturk replied:
But then this is only testing if the code in CLI project is the same as the one in `dotnet-new3.UnitTests` project. It doesn't test if column `12300k` comes before `<1k`. If there is a bug in this comparer and a package with unknown count is always at the top, the test won't be able to catch it. Right?

If that is the case, then maybe we can fix this by writing another test method to test just this class in UnitTests project. 

Or maybe I'm just having its-friday-evening-maybe-you-should-stop-working-for-the-week syndrome, not sure 🙂

vlada-shubina replied:
well, my thinking is as follows:
- if this part is accidentally broken, this piece of code will allow to catch it
- if this part is intentionally changed, this piece of code should also change. 

At the moment, it's not exactly the same as in CLI, and I cannot think of any other approach to verify sorting, do you?

bekir-ozturk replied:
We can verify the sorting by writing a new test method that tests this `DowlnoadCountComparer.Compare` method. This would simply look something like this:
```csharp
[DataRow("12k", "3k", 1)]
[DataRow(null, "", -1)]
[DataRow("<1k", null, 1)]
[DataRow("7k", "7k", 0)]
[DataRow("<1k", "<1k", 0)]
public void TestCompare(string? x, string?y, int expectedOutcome)
{
    Assert.AreEqual(expectedOutcome, new DownloadCountComparer().Compare(x, y));
}
```

vlada-shubina replied:
I added the tests for TotalDownloadsComparer (that now uses long). 
I don't think comparer in tests :) need a test.

DavidKarlas reviewed (approved):
Layouting library depending on strings is a bit sad... Maybe we could expand it to accept object so sorting could be done on integers but since we control string generation sorting by it seems pretty safe, approved.

vlada-shubina commented:
We agreed the following:

- the order as in PR is correct: descending by thousands count, then by name. Alternative for order descending by raw download count was rejected as more confusing.
- the localization of 'k" issue extracted to separate issue to be done in .NET 7. https://github.com/dotnet/templating/issues/4223
- i've updated the PR however to do comparison on long value instead before rendering, Performance wise it's identical whether to do ordering when rendering or before, however comparing longs is cheaper, and added tests for new comparer.
